working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <identifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)* 
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::= 
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*Note that `<type>` derives `<typeIdentifier>`, which makes `<identifier>`
+nearly redundant. However, `<identifier>` is still needed because it
+includes some strings that cannot be the name of a type, e.g., the built-in
+identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
+declared type of a variable or parameter, as the return type of a function
+or getter, as a type argument in a type, or as the on-type of an extension.
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
+where `v` is a fresh variable whose declared type
+is the on-type corresponding to `Ext<S1, .. Sm>`, and similarly for
+instance getters and operators. This rule also applies when a member access
```

Comment:
I'm still a bit confused about where you're talking about extensions vs extension types, but assuming you really mean for this to be about extensions, I think this isn't coherent with the existing resolution rules for extensions.  Specifically, in an extension definition (now), an implicit `this` access always goes to the method on the extension, not to the `on` type method.  The rule you describe here would change that when the `on` type was an extension type, which seems broken.

Reply:
Sorry, I'm not following your reply.  The text that I read here is as follows:

```
If e is an expression whose static type is the extension type Ext<S1, .. Sk>, then a member access like e.m(args) is treated as Ext<S1, .. Sk>(v).m(args) where v is a fresh variable whose declared type is the on-type corresponding to Ext<S1, .. Sm>, and similarly for instance getters and operators. This rule also applies when a member access implicitly has the receiver this, and the static type of this is an extension type (which can only occur in an extension type member declaration).
```

This specifies a semantics which is inconsistent with the existing extension method semantics, right?  So the fix, if there is one, needs to be here.  Or am I misunderstanding something?

Reply:
An "implicit this access" is really a successful lexical scope look-up.

If that lookup finds an extension method declaration, then we don't add a `this.` which is then *implicitly* invoked with an extension method. We convert it *directly* to `Ext<...>(this).method(...)`.

We always look up an unqualified identifier, like `method` in `method(...)`, in the lexical scope. What we then do depends on *what* we find. If it's a static method, it's invoked as `ClassName.method(...)`. If it's a top-level method, it's invoked just as `method(...)`. If it's an instance method it's invoked as `this.method(...)`. And if it's an extension method, it's invoked as `Ext<...>(this).method(...)`. If we find *nothing*, and we are in an instance scope, then we check the interface of the surrounding class, and if we find something there, we do `this.method()`. This is a different step from finding the instance method in the lexical lookup, it just ends up looking the same.

I see no change to the existing specified behavior here.

Reply:
@lrhn I didn't follow that sorry.  I don't think you're responding to what I said?  Let me try to be as explicit as possible. Consider this code:

```dart
extension E1 on int {
  bool get isEven => true;
  void test() { 
      if (isEven) {} // this code is type correct, because implicit this access resolves to the extension
  }
}

extension E2 on E1 {
  List<int> get isEven => [0];
  void test2() {
    isEven[0]; // This code, per the text that I reference here, is invalid
  }
}
```

I claim that the call to isEven in the second extension is an error per this text, because first:

```
If e is an expression whose static type is the extension type Ext<S1, .. Sk>, then a member access like e.m(args) is treated as Ext<S1, .. Sk>(v).m(args) ... This rule also applies when a member access implicitly has the receiver this
```

In the test2, the reference to `isEven` has implicit receiver `this` (so the new rule from @eernstg applies), and so this is treated as `E1(this).isEven`.

This semantics is **inconsistent**.  It is not reasonable to change the method look up in an extension depending on whether or not the `on` type is an extension type or not.

That's my concern.  Am I misunderstanding something?

Reply:
I certainly did not intend to specify a rule which is inconsistent with the current extension methods declaration. However, the case where the on-type is an extension type is already new, so no particular treatment of that case can be a breaking change. Let's consider the example:

```dart
extension E1 on int {
  bool get isEven => true; 
  void test() {
    if (isEven) {} // (1)
  }
}
extension E2 on E1 {
  List<int> get isEven => [0];
  void test2() {
    isEven[0]; // (2)
  }
}
```

At (1), I agree that

> this code is type correct, because implicit this access resolves to the extension

We have an identifier reference `isEven`. Lexical lookup finds the [extension member](https://github.com/dart-lang/language/blob/53f3e4606f7efe25679c37698e167734df7f934c/specification/dartLangSpec.tex#L16094), because it occurs in an enclosing scope, and that extension member has a suitable type.

I actually forgot to specify the identifier reference static/dynamic semantics for this case. PR https://github.com/dart-lang/language/pull/1489 fixes that. I do not expect that specification to be surprising or controversial.

At (2), I don't agree with this:

> This code, per the text that I reference here, is invalid

We have an unqualified function invocation, `isEven[0]`, we perform lexical lookup on `isEven`, we find the declaration `E2.isEven`, and we treat the expression as `E2(this).isEven[0]`. Why would that be invalid?